---
title: "Review Analysis - A tale of native american whole genome sequencing"
author:
- "Developer / Analyst: Israel Aguilar"
- "With Data from: Josue Guzman"
- "Data collection: Aguilar's Virtual Lab Team"
date: "12/19/2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE )

# read libs
library( "pacman" )

p_load( "openxlsx",
        "dplyr",
        "ggplot2",
        "ggrepel",
        "treemapify",
        "cowplot",
        "stringr",
        "ggsci",
        "tidyr",
        "scales",
        "knitr" )
```

# Count projects by technology (Array, WES, WGS)
```{r }
# Cargar datos
allprojects <- read.xlsx( xlsxFile = "Tables A tale of native american whole genome sequencing.xlsx", sheet = "Table 2", startRow = 2 )

# say how many projects are under review
message( nrow(allprojects), " reviewed projetcs" )

# create a DF to summarize projects
bytech <- data.frame(
  technology = c(
    "Array",
    "Whole Exome Sequencing (WES)",
    "Whole Genome Sequencing (WGS)"
  ),
  projects = c( 
    allprojects %>% filter( str_detect( string = Technology,
                                        pattern = "Genotyping" ) ) %>% nrow(),
    allprojects %>% filter( str_detect( string = Technology,
                                        pattern = "WES" ) ) %>% nrow(),
    allprojects %>% filter( str_detect( string = Technology,
                                        pattern = "WGS" ) ) %>% nrow()
  )
)

kable( x = bytech )

```

# Count individuals by technology (Array, WES, WGS)
```{r}
# sum every individual studied
paste( sum( allprojects$WES.Samples,
            allprojects$WGS.Samples,
            allprojects$Arrays.Samples ), "NatAm individuals have been studied" )

# sum by tech
paste( sum( allprojects$WES.Samples), "studied by WES" ) 
paste( sum( allprojects$WGS.Samples), "studied by WGS" ) 
paste( sum( allprojects$Arrays.Samples), "studied by Array tech" ) 
```

# Timeline for projects
```{r}
# Simplify data
# we require:
# project year of publication
# Technology
# Category of samples ( Contemporaneous, ancient, or both )
# WES Samples
# WGS Samples
# Array Samples
# Data.Availability
simplified <- allprojects %>% 
  select( Year.publication,
          Technology,
          `NatAm.DNA.studied.(Contemporaneous,.aDNA,.Both)`, WES.Samples,
          WGS.Samples, Arrays.Samples,
          Data.Availability ) %>% 
  rename( year = Year.publication,
          time_of_samples = `NatAm.DNA.studied.(Contemporaneous,.aDNA,.Both)` )

# extract individual DFs by type of samples
array_projects <- simplified %>% 
  filter( Arrays.Samples > 0 ) %>% 
  select( -WES.Samples, -WGS.Samples ) %>%
  mutate( Technology = "Array" ) %>% 
  rename( samples = Arrays.Samples )

WES_projects <- simplified %>% 
  filter( WES.Samples > 0 ) %>% 
  select( -Arrays.Samples, -WGS.Samples ) %>%
  mutate( Technology = "WES" ) %>% 
  rename( samples = WES.Samples )

WGS_projects <- simplified %>% 
  filter( WGS.Samples > 0 ) %>% 
  select( -Arrays.Samples, -WES.Samples ) %>%
  mutate( Technology = "WGS" ) %>% 
  rename( samples = WGS.Samples )

# gather all techs
alltech <- bind_rows( array_projects,
                      WES_projects,
                      WGS_projects ) %>%
  mutate( sample_tag = case_when( samples < 10 ~ "< 10",
                                  samples < 100 ~ "< 100",
                                  samples < 1000 ~ "< 1000",
                                  TRUE ~ "more than 1000"
  ) )
```

```{r}
# lets try dots
timeline <- ggplot( data = alltech,
                    mapping = aes( x = year,
                                   y = Technology ) ) +
  geom_point( mapping = aes(size = sample_tag,
                            fill = Data.Availability) ,
              shape = 21, alpha = 0.5,
              position = position_jitter( width = 0,
                                          height = 0.2,
                                          seed = 66 ) ) +
  scale_x_continuous( limits = c(2009, 2021),
                      breaks = seq( 2009, 2021, 2) ) +
  scale_fill_startrek( ) +
  labs( title = "Timeline for Native American genomic projects",
        x = "year of publication",
        y = "technology used",
        fill = "Data Available?",
        size = "n. of studied indiviuals" ) +
  theme_light( base_size = 10 ) +
  theme( panel.grid.major.x = element_line( linetype = "dashed" ),
         panel.grid.minor.x = element_line( linetype = "dashed" ),
         axis.title = element_blank( ),
         axis.text.y = element_text( face = "bold"),
         plot.title = element_text( face = "bold" ),
         legend.title = element_text( face = "bold" ) )

# save plot for manuscript
ggsave( filename = "timeline.png",
        plot = timeline,
        width = 7,
        height = 5,
        dpi = 600 )

```
# Accumulated number of individuals studied

```{r}
# line plot by tech
# number of samples acuumulated
sum_by_year <- alltech %>% 
  group_by( year, Technology ) %>% 
  summarize( total_year = sum(samples) ) %>% 
  ungroup( ) %>% 
  group_by( Technology ) %>% 
  summarize( acc = cumsum(total_year),
             year = year,
             total_year = total_year )

# plot
# prepare for y axis control
thebreaks <- seq( 0, 9000, 1000 )

line_dot <- ggplot( data = sum_by_year,
                    mapping = aes( x = year,
                                   y = acc,
                                   group = Technology,
                                   color = Technology,
                                   # lty = Technology
                    ) ) +
  geom_line( size = 0.4 ) +
  geom_point( data = sum_by_year %>% 
                filter( year == 2021 ) ) +
  scale_y_continuous( position = "right",
                      breaks = thebreaks,
                      labels = prettyNum( thebreaks, big.mark = "," ) ) +
  scale_x_continuous( limits = c(2009, 2021),
                      breaks = seq( 2009, 2021, 2) ) +
  scale_color_jama( ) +
  labs( title = "The increase in NatAm sequencing",
        x = "year of publishing",
        y = "Acc. studied individuals") +
  theme_classic( ) +
  theme( legend.position = "top",
         axis.title = element_text( face = "bold" ),
         legend.title = element_text( face = "bold" ),
         plot.title = element_blank( ) )

# save plot for manuscript
ggsave( filename = "acc_individuals.png",
        width = 7,
        height = 5,
        dpi = 600 )
```

```{r}
# lets try a panel fig
plot_grid( timeline,
           line_dot,
           ncol = 1, align = "v",
           axis = "lr" )

# save plot for manuscript
ggsave( filename = "timepanel.png",
        width = 7,
        height = 10,
        dpi = 600 )
```

# Claculate the share of available samples

```{r }
full_natam_samples <- read.xlsx( xlsxFile = "Tables A tale of native american whole genome sequencing.xlsx", sheet = "Sup Tabl 2.", startRow = 2 ) %>% 
  mutate( year = as.numeric( Publication_Year ) )

# clear the df and summarize
cleared <- full_natam_samples %>% 
  select( year, Technology, Data_availability ) %>% 
  group_by( year, Technology, Data_availability ) %>% 
  summarise( bydata_total_year = n( ) )
```

```{r }
# lets create the base dataframe
all_years <- unique( cleared$year ) %>% sort( )

all_technologies <- unique( cleared$Technology ) %>% sort( )

all_availability <- unique( cleared$Data_availability ) %>% sort( )

# create the crossinf of variables
base_df <- crossing( all_years, all_technologies, all_availability ) %>% 
  rename( year = all_years,
          Technology = all_technologies,
          Data_availability = all_availability )

# annotate base df
annotated <- left_join( x = base_df,
                        y = cleared,
                        by = c( "year" = "year",
                                "Technology" = "Technology",
                                "Data_availability" = "Data_availability" ) ) %>% 
  mutate( bydata_total_year = ifelse( test = is.na( bydata_total_year ), yes = 0, no = bydata_total_year ) )

# create a function for stacked area plot
plot_areas <- function( the_data, the_tech ){
  
  ## lets try only with arrays
  percentages <- the_data %>%
    filter( Technology == the_tech ) %>% 
    group_by( Data_availability ) %>%
    summarize( year = year,
               Technology = Technology,
               bydata_total_year = bydata_total_year,
               accumulated_bydata = cumsum( bydata_total_year ) ) %>% 
    ungroup( ) %>% 
    group_by( year ) %>% 
    summarize( Data_availability,
               Technology = Technology,
               bydata_total_year = bydata_total_year,
               accumulated_bydata = accumulated_bydata,
               total_accumulated_bydata_peryear = sum(accumulated_bydata),
               fraction = accumulated_bydata / total_accumulated_bydata_peryear ) %>% 
    ungroup( ) %>% 
    mutate( fraction = ifelse( test = is.nan( fraction ),
                               yes = 0,
                               no = fraction ) )
  
  # show table for 2021 share of available data
  print( the_tech )
  print( kable( x = percentages %>% filter( year == 2021 ) %>% select(year, Data_availability, accumulated_bydata, fraction ) ) )
  
  # Create area plot...
  ggplot( data = percentages,
          mapping = aes( x = year,
                         y = fraction,
                         fill = Data_availability ) ) + 
    geom_area( alpha = 0.6,
               size = 1,
               colour = "black" ) +
    scale_x_continuous( limits = c(2009, 2021),
                        breaks = seq( 2009, 2021, 1),
                        expand = c( 0, 0 ) ) +
    scale_y_continuous( labels = percent, 
                        expand = c(0, 0),
                        sec.axis = sec_axis( trans = ~.,
                                             name = "accumulated data",
                                             labels = percent ) ) +
    # scale_fill_startrek( ) +
    scale_fill_manual( values = c("#cc1c00", "#84bd00")) +
    labs( title = "Trends in Data Availability",
          y = the_tech,
          fill = "Data available?" ) +
    theme_light( ) +
    theme( panel.grid.minor = element_blank( ),
           axis.text.y.left = element_blank( ),
           axis.ticks.y.left = element_blank( ),
           panel.background = element_rect( color = "black",
                                            size = 2, linetype = "solid" ) )
  
}

# plot the DATA!

wgs_stack <- plot_areas( the_data = annotated,
                         the_tech = "WGS"  )

wes_stack <-plot_areas( the_data = annotated,
                        the_tech = "WES"  )

array_stack <- plot_areas( the_data = annotated,
                           the_tech = "Arrays"  )

# create a panel
plot_grid( wgs_stack +
             theme( axis.title.x = element_blank( ),
                    axis.text.x = element_blank( ),
                    legend.position = "none" ),
           wes_stack +
             theme( axis.title.x = element_blank( ),
                    axis.text.x = element_blank( ),
                    plot.title = element_blank( ) ),
           array_stack +
             theme( plot.title = element_blank( ),
                    axis.text.x = element_text( angle = 90, vjust = 0.5, hjust = 0.5 ),
                    legend.position = "none" ),
           ncol = 1,
           axis = "lr",
           align = "v",
           rel_heights = c( 0.35, 0.3, 0.35)  )

# save plot for manuscript
ggsave( filename = "sharepanel.png",
        width = 7,
        height = 10,
        dpi = 600 )

# print total samples with availble data
message( annotated %>%
           filter(Data_availability == "Yes") %>%
           pull(bydata_total_year) %>% 
           sum( ),
         " samples with available data")
```

## The openness in ancient studies
```{r}
##  get the number of ancient samples
message( full_natam_samples %>%
           filter( DNA_type_of_samples == "AncientDNA" ) %>% 
           nrow( ) ,
         " samples are ancient DNA")

# Lets create a pie chart for ancient and modern dna
pie_data <- full_natam_samples %>% 
  group_by( DNA_type_of_samples ) %>% 
  summarise( samples = n( ) ) %>% 
  mutate( fraction = samples / sum(samples) ,
          percent = percent(fraction),
          lab_pos = 1 - ( fraction / 2) ) %>% 
  ungroup( )

# print table
kable( x = pie_data )

# plot pie...
pieplot <- ggplot( data = pie_data,
                   mapping = aes( x = 2,
                                  y = fraction,
                                  fill = DNA_type_of_samples ) ) +
  geom_label_repel( mapping = aes( label = paste( DNA_type_of_samples,
                                                  "\n",
                                                  prettyNum( samples, big.mark = "," ),
                                                  "samples" ),
                                   x = 2,
                                   y = lab_pos ),
                    min.segment.length = 0,
                    segment.curvature = -1e-20,
                    nudge_x = 1,
                    fill = "white" ) +
  geom_col( color = "black" ) +
  xlim( c(0, 3) ) +
  scale_fill_manual( values = c( 
    "deepskyblue3",
    "cornflowerblue") ) +
  coord_polar( theta = "y" ) +
  labs( title = "Ancient DNA in NatAm genomics" ) +
  theme_void( ) +
  theme( legend.position = "none",
         plot.background = element_rect( fill = "white") )

```

```{r}

# Get the share of available data per type of DNA
fractions <- full_natam_samples %>% 
  group_by( DNA_type_of_samples,
            Data_availability ) %>% 
  summarise( samples = n( ) ) %>% 
  ungroup(  ) %>% 
  group_by( DNA_type_of_samples ) %>% 
  summarise( Data_availability = Data_availability,
             samples = samples,
             totals = sum(samples),
             frac = samples / totals )

# print table
kable( x = fractions )

# Lets plot cols stacked
sharebar <- ggplot( data = fractions,
                    mapping = aes( x = DNA_type_of_samples,
                                   y = frac,
                                   fill = Data_availability ) ) +
  geom_col( color = "black", width = 0.9 ) +
  scale_fill_manual( values = c("#cc1c00", "#84bd00")) +
  scale_y_continuous( limits = c(0,1),
                      breaks = seq( 0, 1, by = 0.2 ),
                      labels = percent,
                      expand = c(0,0) ) +
  labs( title = "Data availability",
        x = "",
        y = "Accumulated samples",
        fill = "Is Publicly Available?" ) +
  theme_classic( ) +
  theme( legend.position = "right",
         axis.text.x = element_text( face = "bold" ) )

# create a panel for ancient overview
ancient_panel <- plot_grid( pieplot,
                            sharebar,
                            nrow = 1,
                            rel_widths = c(0.35, 0.65) )

```

```{r }
# create a function for treemap
treemaper <- function( the_data, the_time, n_cut, the_color ) {
  #
  tmp_data <- the_data %>% 
    filter( DNA_type_of_samples == the_time )
  
  # sample cutofffor other tagging
  sample_cutoff <- n_cut
  
  # Lets create a treemap to see the most studied pops
  pop_summary <- tmp_data %>% 
    group_by( Population ) %>% 
    summarise( samples = n( ) ) %>% 
    arrange( -samples ) %>% 
    ungroup( ) %>% 
    mutate( tag = ifelse( test = samples < sample_cutoff,
                          yes = "other",
                          no = Population )) %>% 
    group_by( tag ) %>% 
    summarise( totals = sum(samples)  ) %>% 
    arrange( -totals ) %>% 
    mutate( units = "individuals" ) 
  
  # number of other pops
  other_n <- tmp_data %>% 
    group_by( Population ) %>% 
    summarise( samples = n( ) ) %>% 
    arrange( -samples ) %>% 
    ungroup( ) %>% 
    filter( samples < sample_cutoff ) %>% 
    nrow( )
  
  #
  pop_summary$units[ 2:nrow(pop_summary) ] <- ""
  
  pop_summary <- pop_summary %>% 
    mutate( subtag = paste( prettyNum( totals, big.mark = ","),
                            units ),
            tag = ifelse( test = tag == "other",
                          yes = paste( "other", other_n, "Pops",
                                       "with less than", sample_cutoff, "individuals"),
                          no = tag ) )
  
  # lets make treemap
  ggplot( data = pop_summary,                     # el datframe que tiene las etiquetas y subetiquetas
          mapping = aes( area = totals ) ) +   # el tamanio de los cuadrados sera proporcional a la cantidad de anomalias contadas
    geom_treemap( 
      mapping = aes(fill = totals ),        # Usamos geometria de treemap; el relleno fill sera un gradiente dependiendo de la cantidad de anomalias
                  color = "black",                          # el borde de los cuadros sera blanco
                  size = 2 ) +                                # el grosor del borde sera de 3 unidades
    geom_treemap_text( mapping = aes( label = tag ),    # su mapeo label indica que columna aparecera en el texto
                       place = "centre",                     # en que parte del cuadro aparecera la palabra
                       reflow = TRUE,                        # ajustar el texto a la caja
                       size = 10,                            # el tamanio de la palabra
                       fontface = "bold" ) +
    geom_treemap_text( mapping = aes( label = subtag ),     # el texto viene de la subetiqueta que creamos
                       place = "bottomright",                    # aparecera abajo a la derecha
                       color = "black",                          # el color del texto es blanco
                       padding.x = unit( x = 1,                  # se puede mover la etiqueta, esta se mueve 3
                                         units = "mm" ),         # milimetros
                       padding.y = unit( x = 1,                  # se mueve en y 3 mm tambien
                                         units = "mm" ),
                       size = 8 ) +
    scale_fill_gradient( low = "white",      # el color mas bajo es este hexadecimal que es un naranjita
                         high = the_color ) +
    labs( title = "Most Studied NatAm populations",
          subtitle = the_time ) +
    theme( legend.position = "none",                       # eliminamos la leyenda; no dice mucho
           plot.title =  element_text( size = 17,          # cambiamos el tamanio del titulo
                                       face = "bold" ) )
}

```



```{r }
modern_tree <- treemaper( the_data = full_natam_samples,
                          the_time = "ModernDNA",
                          the_color = "cornflowerblue",
                          n_cut = 100 )

ancient_tree <- treemaper( the_data = full_natam_samples,
                           the_time = "AncientDNA",
                           the_color = "deepskyblue3",
                           n_cut = 5 )

treegrid <- plot_grid( modern_tree,
                       ancient_tree + theme( plot.title = element_blank( ) ),
                       nrow = 1,
                       align = "h" )

# gather grids
plot_grid( ancient_panel,
           treegrid,
           ncol = 1,
           rel_heights = c( 0.4, 0.6 ) )

# save plot for manuscript
ggsave( filename = "ancientpanel.png",
        width = 10,
        height = 10,
        dpi = 600 )

```