---
title: "Review Analysis - A tale of native american whole genome sequencing"
author:
- "Developer / Analyst: Israel Aguilar"
- "With Data from: Josue Guzman"
- "Data collection: Aguilar's Virtual Lab Team"
date: "12/19/2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE )

# read libs
library( "pacman" )

p_load( "openxlsx",
        "dplyr",
        "ggplot2",
        "cowplot",
        "stringr",
        "ggsci",
        "tidyr",
        "scales",
        "knitr" )
```

# Count projects by technology (Array, WES, WGS)
```{r }
# Cargar datos
allprojects <- read.xlsx( xlsxFile = "Tables A tale of native american whole genome sequencing.xlsx", sheet = "Table 2", startRow = 2 )

# say how many projects are under review
message( nrow(allprojects), " reviewed projetcs" )

# create a DF to summarize projects
bytech <- data.frame(
  technology = c(
    "Array",
    "Whole Exome Sequencing (WES)",
    "Whole Genome Sequencing (WGS)"
  ),
  projects = c( 
    allprojects %>% filter( str_detect( string = Technology,
                                        pattern = "Genotyping" ) ) %>% nrow(),
    allprojects %>% filter( str_detect( string = Technology,
                                        pattern = "WES" ) ) %>% nrow(),
    allprojects %>% filter( str_detect( string = Technology,
                                        pattern = "WGS" ) ) %>% nrow()
  )
)

kable( x = bytech )

```

# Count individuals by technology (Array, WES, WGS)
```{r}
# sum every individual studied
paste( sum( allprojects$WES.Samples,
            allprojects$WGS.Samples,
            allprojects$Arrays.Samples ), "NatAm individuals have been studied" )

# sum by tech
paste( sum( allprojects$WES.Samples), "studied by WES" ) 
paste( sum( allprojects$WGS.Samples), "studied by WGS" ) 
paste( sum( allprojects$Arrays.Samples), "studied by Array tech" ) 
```

# Timeline for projects
```{r}
# Simplify data
# we require:
# project year of publication
# Technology
# Category of samples ( Contemporaneous, ancient, or both )
# WES Samples
# WGS Samples
# Array Samples
# Data.Availability
simplified <- allprojects %>% 
  select( Year.publication,
          Technology,
          `NatAm.DNA.studied.(Contemporaneous,.aDNA,.Both)`, WES.Samples,
          WGS.Samples, Arrays.Samples,
          Data.Availability ) %>% 
  rename( year = Year.publication,
          time_of_samples = `NatAm.DNA.studied.(Contemporaneous,.aDNA,.Both)` )

# extract individual DFs by type of samples
array_projects <- simplified %>% 
  filter( Arrays.Samples > 0 ) %>% 
  select( -WES.Samples, -WGS.Samples ) %>%
  mutate( Technology = "Array" ) %>% 
  rename( samples = Arrays.Samples )

WES_projects <- simplified %>% 
  filter( WES.Samples > 0 ) %>% 
  select( -Arrays.Samples, -WGS.Samples ) %>%
  mutate( Technology = "WES" ) %>% 
  rename( samples = WES.Samples )

WGS_projects <- simplified %>% 
  filter( WGS.Samples > 0 ) %>% 
  select( -Arrays.Samples, -WES.Samples ) %>%
  mutate( Technology = "WGS" ) %>% 
  rename( samples = WGS.Samples )

# gather all techs
alltech <- bind_rows( array_projects,
                      WES_projects,
                      WGS_projects ) %>%
  mutate( sample_tag = case_when( samples < 10 ~ "< 10",
                                  samples < 100 ~ "< 100",
                                  samples < 1000 ~ "< 1000",
                                  TRUE ~ "more than 1000"
  ) )
```

```{r}
# lets try dots
timeline <- ggplot( data = alltech,
                    mapping = aes( x = year,
                                   y = Technology ) ) +
  geom_point( mapping = aes(size = sample_tag,
                            fill = Data.Availability) ,
              shape = 21, alpha = 0.5,
              position = position_jitter( width = 0,
                                          height = 0.2,
                                          seed = 66 ) ) +
  scale_x_continuous( limits = c(2009, 2021),
                      breaks = seq( 2009, 2021, 2) ) +
  scale_fill_startrek( ) +
  labs( title = "Timeline for Native American genomic projects",
        x = "year of publication",
        y = "technology used",
        fill = "Data Available?",
        size = "n. of studied indiviuals" ) +
  theme_light( base_size = 10 ) +
  theme( panel.grid.major.x = element_line( linetype = "dashed" ),
         panel.grid.minor.x = element_line( linetype = "dashed" ),
         axis.title = element_blank( ),
         axis.text.y = element_text( face = "bold"),
         plot.title = element_text( face = "bold" ),
         legend.title = element_text( face = "bold" ) )

# save plot for manuscript
ggsave( filename = "timeline.png",
        plot = timeline,
        width = 7,
        height = 5,
        dpi = 600 )

```
# Accumulated number of individuals studied

```{r}
# line plot by tech
# number of samples acuumulated
sum_by_year <- alltech %>% 
  group_by( year, Technology ) %>% 
  summarize( total_year = sum(samples) ) %>% 
  ungroup( ) %>% 
  group_by( Technology ) %>% 
  summarize( acc = cumsum(total_year),
             year = year,
             total_year = total_year )

# plot
# prepare for y axis control
thebreaks <- seq( 0, 9000, 1000 )

line_dot <- ggplot( data = sum_by_year,
                    mapping = aes( x = year,
                                   y = acc,
                                   group = Technology,
                                   color = Technology,
                                   # lty = Technology
                    ) ) +
  geom_line( size = 0.4 ) +
  geom_point( data = sum_by_year %>% 
                filter( year == 2021 ) ) +
  scale_y_continuous( position = "right",
                      breaks = thebreaks,
                      labels = prettyNum( thebreaks, big.mark = "," ) ) +
  scale_x_continuous( limits = c(2009, 2021),
                      breaks = seq( 2009, 2021, 2) ) +
  scale_color_jama( ) +
  labs( title = "The increase in NatAm sequencing",
        x = "year of publishing",
        y = "Acc. studied individuals") +
  theme_classic( ) +
  theme( legend.position = "top",
         axis.title = element_text( face = "bold" ),
         legend.title = element_text( face = "bold" ),
         plot.title = element_blank( ) )

# save plot for manuscript
ggsave( filename = "acc_individuals.png",
        width = 7,
        height = 5,
        dpi = 600 )
```

```{r}
# lets try a panel fig
plot_grid( timeline,
           line_dot,
           ncol = 1, align = "v",
           axis = "lr" )

# save plot for manuscript
ggsave( filename = "timepanel.png",
        width = 7,
        height = 10,
        dpi = 600 )
```

# Claculate the share of available samples

```{r }
full_natam_samples <- read.xlsx( xlsxFile = "Tables A tale of native american whole genome sequencing.xlsx", sheet = "Sup Tabl 2.", startRow = 2 ) %>% 
  mutate( year = as.numeric( Publication_Year ) )

# clear the df and summarize
cleared <- full_natam_samples %>% 
  select( year, Technology, Data_availability ) %>% 
  group_by( year, Technology, Data_availability ) %>% 
  summarise( bydata_total_year = n( ) )
```

```{r }
# lets create the base dataframe
all_years <- unique( cleared$year ) %>% sort( )

all_technologies <- unique( cleared$Technology ) %>% sort( )

all_availability <- unique( cleared$Data_availability ) %>% sort( )

# create the crossinf of variables
base_df <- crossing( all_years, all_technologies, all_availability ) %>% 
  rename( year = all_years,
          Technology = all_technologies,
          Data_availability = all_availability )

# annotate base df
annotated <- left_join( x = base_df,
                        y = cleared,
                        by = c( "year" = "year",
                                "Technology" = "Technology",
                                "Data_availability" = "Data_availability" ) ) %>% 
  mutate( bydata_total_year = ifelse( test = is.na( bydata_total_year ), yes = 0, no = bydata_total_year ) )

# create a function for stacked area plot
plot_areas <- function( the_data, the_tech ){
  
  ## lets try only with arrays
  percentages <- the_data %>%
    filter( Technology == the_tech ) %>% 
    group_by( Data_availability ) %>%
    summarize( year = year,
               Technology = Technology,
               bydata_total_year = bydata_total_year,
               accumulated_bydata = cumsum( bydata_total_year ) ) %>% 
    ungroup( ) %>% 
    group_by( year ) %>% 
    summarize( Data_availability,
               Technology = Technology,
               bydata_total_year = bydata_total_year,
               accumulated_bydata = accumulated_bydata,
               total_accumulated_bydata_peryear = sum(accumulated_bydata),
               fraction = accumulated_bydata / total_accumulated_bydata_peryear ) %>% 
    ungroup( ) %>% 
    mutate( fraction = ifelse( test = is.nan( fraction ),
                               yes = 0,
                               no = fraction ) )
  
  # show table for 2021 share of available data
  print( the_tech )
  print( kable( x = percentages %>% filter( year == 2021 ) %>% select(year, Data_availability, accumulated_bydata, fraction ) ) )
  
  # Create area plot...
  ggplot( data = percentages,
          mapping = aes( x = year,
                         y = fraction,
                         fill = Data_availability ) ) + 
    geom_area( alpha = 0.6,
               size = 1,
               colour = "black" ) +
    scale_x_continuous( limits = c(2009, 2021),
                        breaks = seq( 2009, 2021, 1),
                        expand = c( 0, 0 ) ) +
    scale_y_continuous( labels = percent, 
                        expand = c(0, 0),
                        sec.axis = sec_axis( trans = ~.,
                                             name = "accumulated data",
                                             labels = percent ) ) +
    # scale_fill_startrek( ) +
    scale_fill_manual( values = c("#cc1c00", "#84bd00")) +
    labs( title = "Trends in Data Availability",
          y = the_tech,
          fill = "Data available?" ) +
    theme_light( ) +
    theme( panel.grid.minor = element_blank( ),
           axis.text.y.left = element_blank( ),
           axis.ticks.y.left = element_blank( ),
           panel.background = element_rect( color = "black",
                                            size = 2, linetype = "solid" ) )
  
}

# plot the DATA!

wgs_stack <- plot_areas( the_data = annotated,
                         the_tech = "WGS"  )

wes_stack <-plot_areas( the_data = annotated,
                        the_tech = "WES"  )

array_stack <- plot_areas( the_data = annotated,
                           the_tech = "Arrays"  )

# create a panel
plot_grid( wgs_stack +
             theme( axis.title.x = element_blank( ),
                    axis.text.x = element_blank( ),
                    legend.position = "none" ),
           wes_stack +
             theme( axis.title.x = element_blank( ),
                    axis.text.x = element_blank( ),
                    plot.title = element_blank( ) ),
           array_stack +
             theme( plot.title = element_blank( ),
                    axis.text.x = element_text( angle = 90, vjust = 0.5, hjust = 0.5 ),
                    legend.position = "none" ),
           ncol = 1,
           axis = "lr",
           align = "v",
           rel_heights = c( 0.35, 0.3, 0.35)  )

# save plot for manuscript
ggsave( filename = "sharepanel.png",
        width = 7,
        height = 10,
        dpi = 600 )

# print total samples with availble data
message( annotated %>%
           filter(Data_availability == "Yes") %>%
           pull(bydata_total_year) %>% 
           sum( ),
         " with available data")
```